lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -822,15 +876,64 @@ SessionConfig SessionState::get_config() const {
   return config_;
 }
 
-uint32_t SessionState::get_local_teid() const {
-  return local_teid_;
-}
-
-void SessionState::set_local_teid(
-    uint32_t teid, SessionStateUpdateCriteria& uc) {
-  local_teid_              = teid;
-  uc.is_local_teid_updated = true;
-  uc.local_teid_updated    = teid;
+uint32_t SessionState::get_upf_local_teid() const {
+  char teidar[10];
+  union mix_t {
+    std::uint32_t theDWord;
+    std::uint8_t theBytes[4];
+  };
+  mix_t aVal;
+  // Retreive value
+  std::string teids = config_.rat_specific_context.m5gsm_session_context()
+                          .upf_endpoint()
+                          .teid();
+  uint32_t len = teids.length();
+  memset(teidar, '\0', 10);
+  memcpy(&teidar[0], &teids[0], len);
+  bool byte_zero_skip = true;
+  for (int i = 0; i < 4; ++i) {
+    if (byte_zero_skip && (!teidar[i])) continue;
+    // Rest continue;
+    byte_zero_skip   = false;
+    aVal.theBytes[i] = uint32_t(teidar[i]);
+  }
+  return ntohl(aVal.theDWord);
+}
+
+void SessionState::set_upf_teid_endpoint(
+    std::string ip_addr, uint32_t teid, SessionStateUpdateCriteria& uc) {
+  char teidar[5];
+  in_addr upfip;
+  // Over right UPF teid details
+  // Lets set the TEID first
+  memset(teidar, '\0', 5);
+  union mix_t {
+    std::uint32_t theDWord;
+    std::uint8_t theBytes[4];
+  };
+  mix_t aVal;
+  aVal.theDWord = teid;
+  aVal.theDWord = htonl(aVal.theDWord);
+  for (int i = 0; i < 4; ++i) {
+    teidar[i] = uint8_t(aVal.theBytes[i]);
+  }
+  config_.rat_specific_context.mutable_m5gsm_session_context()
+      ->mutable_upf_endpoint()
+      ->set_teid(teidar);
+
+  // Lets set the UPF endpoint ip address
+  memset(teidar, '\0', 5);
+  inet_aton(ip_addr.c_str(), &upfip);
+  for (int i = 3; i >= 0; i--) {
+    teidar[i] = (upfip.s_addr & 0x000000FF);
+    upfip.s_addr >>= 8;
+  }
+  // Let set UPF ip first
+  config_.rat_specific_context.mutable_m5gsm_session_context()
+      ->mutable_upf_endpoint()
+      ->set_end_ipv4_addr(teidar);
+  uc.updated_config    = config_;
+  uc.is_config_updated = true;
```

Comment:
I may be missing something, but isn't it an easier way to do this conversion.  Usage of memset is always dangerous and can introduce bugs difficult to troubleshoot

Reply:
Why aren't we just storing the teid as and uint32_t and let the upf do the conversion?

Reply:
As resonss mentiond above why we can't used in uin32_t format, we used with right length checks, so there shud't be any issues.